### PR TITLE
create NNlibCUDA sub-package

### PR DESCRIPTION
**Pull request by CarloLucibello**
Follow up to discussion in https://github.com/JuliaGPU/CUDA.jl/pull/738 and on zulip. 
The idea is to create a package within the same repo of NNlib.jl to make maintenance easier.

This a port and factorization of the nnlib.jl file and related in CUDA.jl, so that CUDA.jl can drop the NNlib dependence (pr JuliaGPU/CUDA.jl#753)

Besides this PR, Flux.jl should remove the dependence on `CUDA.batchnorm` and use the new cudnn wrapper interface.

The CI pipeline is copied from https://github.com/JuliaGPU/KernelAbstractions.jl

cc @maleadt @denizyuret

THINGS TO DO AFTER MERGE:
- Test Flux against this and CUDA master
- Tag a release
- Benchmark softmax, see if we want to keep the cudnn version or the native one. Same for batchnorm.

### CODE REVIEW

File: lib/NNlibCUDA/test/upsample.jl [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+            7//5 26//15 31//15 12//5;
+            9//5 32//15 37//15 14//5;
+           11//5 38//15 43//15 16//5;
+           13//5 44//15 49//15 18//5;
+            3//1 10//3  11//3   4//1]
+    y_true = cat(y_true,y_true; dims=3)
+    y_true = cat(y_true,y_true; dims=4)
+    y_true = cu(y_true)
+  
+    @allowscalar y = upsample_bilinear(x, (3,2))
```
Comment (CarloLucibello) [inline lib/NNlibCUDA/test/upsample.jl:17]: @maxfreu this would error out when `allowscalar(false)`, is that normal?

File: lib/NNlibCUDA/test/activations.jl [hunk 2]
```diff
@@ -0,0 +1,5 @@
+@testset "Broadcast Fix" begin
+    if CUDA.has_cudnn()
+        @test testf(x -> logσ.(x), rand(5))
```
Comment (CarloLucibello) [inline lib/NNlibCUDA/test/activations.jl:3]: I don't know what this `ftest` is doing, would be good to have some internal testing utils instead

File: lib/NNlibCUDA/test/upsample.jl (hunk 1 above)
Comment (DhairyaLGandhi) [reply lib/NNlibCUDA/test/upsample.jl:17]: Please mark that as broken for rhe time being.

Comment (CarloLucibello) [reply lib/NNlibCUDA/test/upsample.jl:17]: I'm not sure this is broken, maybe it is just the way it is and scalar indexing is not impacting performance in this case. @maxfreu will know more since he implemented it

Comment (DhairyaLGandhi) [reply lib/NNlibCUDA/test/upsample.jl:17]: Since our requirement is for the suite to not perform scalar indexing, and it errors in that state, I think it's fair to say something needs correcting. We can investigate the cause for that later, and good to intimate the author.

**Comment by CarloLucibello** [comment]
@DhairyaLGandhi could you add buildkite to the repo and get rid of AppVeyor?

File: lib/NNlibCUDA/test/upsample.jl (hunk 1 above)
Comment (maxfreu) [reply lib/NNlibCUDA/test/upsample.jl]: You have encountered a problem this very PR is going to fix: You have to be on CUDA#master, since there has not been a release including the underlying functions yet. So here it falls back to the CPU implementation, since it finds nothing better.

Comment (CarloLucibello) [reply lib/NNlibCUDA/test/upsample.jl]: my bad, I wasn't using the kernel I just ported, fixed now

**Comment by CarloLucibello** [comment]
Tests are passing locally on top of JuliaGPU/CUDA.jl#738. This is pretty much done. Before merging, we should test it against JuliaGPU/CUDA.jl#738 (once it is merged in master) using buildkite  (bump on @DhairyaLGandhi for activating it)

**Comment by CarloLucibello** [comment]
CUDA master removed NNlib, we should hurry up with this. @DhairyaLGandhi can you activate buildkite for this repo?

**Comment by maleadt** [comment]
https://buildkite.com/julialang/nnlib-dot-jl

File: .ci/test_nnlibcuda.jl [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import Pkg
+# import NNlibCUDA
+pkgs = ["NNlibCUDA"]
+
+# Do this for the time being since test doesn't pick up the manifest
+# for some reason. Can remove this and manifests when cuda 3.0 is released. 
+Pkg.add(url="https://github.com/JuliaGPU/CUDA.jl.git", rev="master")
```
Comment (CarloLucibello) [inline .ci/test_nnlibcuda.jl:7]: @maleadt I'm trying to point to CUDA master but I'm having no luck. This line is working for me executing the script locally but doesn't work with buildkite. Do you have any suggestions?

Comment (CarloLucibello) [reply .ci/test_nnlibcuda.jl:7]: Also the Manifest seems to be ignored when executing this script

Comment (maleadt) [reply .ci/test_nnlibcuda.jl:7]: It doesn't get to that but fails during precompilation, after installing the package. The only solution is to add a Manifest, I guess.

Comment (CarloLucibello) [reply .ci/test_nnlibcuda.jl]: I tried locally but `test` didn't see the Manifest, don't know why. uploaded it here, let's see

Comment (maleadt) [reply .ci/test_nnlibcuda.jl]: CUDA.jl also has the manifest committed, and that works.

Comment (CarloLucibello) [reply .ci/test_nnlibcuda.jl]: I don't know why it doesn't work here, anyways I worked around moving the Pkg.add before the precompile. 
Can be revisited later
